Throw more specific / PSR-11 exceptions

skalpa opened the pull request:
Pimple now throws more specific exceptions.

#### An attempt to modify a frozen service was made
- Before: `\RuntimeException`
- After: `FrozenServiceException extends \RuntimeException`

#### The identifier of a valid service or parameter was expected
- Before: `\InvalidArgumentException`
- After: `UnknownIdentifierException extends \InvalidArgumentException`

#### A closure or invokable object was expected
- Before: `\InvalidArgumentException`
- After: `ExpectedInvokableException extends \InvalidArgumentException`

#### A attempt to perform an operation that requires a service identifier was made
Thrown when trying to `extend()` a parameter

- Before: `\InvalidArgumentException`
- After: `InvalidServiceIdentifierException extends \InvalidArgumentException`

## PSR-11 compatibility

All the classes implement `Psr\Container\ContainerExceptionInterface` and `UnknownIdentifierException` implements `NotFoundExceptionInterface`.

I believe it doesn't hurt and couldn't find a reason not to do it like this.

Also, this will make PSR-11 wrappers faster. Until now, these wrappers have to do something like this:

```php
class PimpleContainer implements \Psr\Container\ContainerInterface
{
    public function get($id)
    {
        if (!isset($this->container[$id]) {
            throw new NotFoundException(sprintf('The identifier %s is not defined.', $id));
        }

        return $this->container[$id];
    }
}
```

The additional cost is: 
- 2 function calls: `ContainerInterface::get()` and `Pimple\Container::offsetExists()`
- 1 `isset()` (the one in `offsetExists()`)

By making the Pimple exceptions Compliant with PSR-11, this cost can be reduced to a single function call:

```php
class PimpleContainer implements \Psr\Container\ContainerInterface
{
    public function get($id)
    {
        return $this->container[$id];
    }
}
```

## Code changes

All the changes are pretty straightforward. The only one that may have to be discussed is the introduction of a test that throws a `FrozenServiceException` in `extend()`. I'll post a separate issue to explain the behavior this fixes.

## The tests

I've kept the old tests and renamed them `legacySomething()`, so we could check that the changes didn't break BC. It's probably safer to keep them, but if you think that's superfluous I can remove them before merge.

fabpot commented on src/Pimple/Exception/InvalidServiceIdentifierException.php:
```diff
@@ -0,0 +1,47 @@
+use Psr\Container\NotFoundExceptionInterface;
+
+/**
+ * A attempt to perform an operation that requires a service identifier was made.
```
An?

fabpot commented on src/Pimple/Exception/InvalidServiceIdentifierException.php:
```diff
@@ -0,0 +1,47 @@
+class InvalidServiceIdentifierException extends \InvalidArgumentException implements NotFoundExceptionInterface
+{
+    /**
+     * Constructor.
```
Can be removed

skalpa replied:
What do you mean ? The line, docblock, or the class ?

fabpot replied:
Sorry, I was not clear enough. I try to avoid unhelful/unneeded phpdocs nowadays. So, here `Contructor` does not add any useful information, so it can be removed.

fabpot commented on src/Pimple/Exception/UnknownIdentifierException.php:
```diff
@@ -0,0 +1,47 @@
+/**
+ * The identifier of a valid service or parameter was expected.
+ *
+ * @author  Pascal Luna <skalpa@zetareticuli.org>
```
Extra space after `@author` (same for the other new files as well)

fabpot commented:
Also, the C extension must be updated as well.

stof commented:
Implementing PSR-11 breaks BC: it makes Pimple incompatible with Silex due to a conflict on the `get` method. This has already been discussed in at least 10 previous PR.

skalpa commented:
@stof Yes I know, I saw some of them, but this is not what this PR does. 

It just adds pimple-specific exception classes that implement `Psr\Container\ExceptionInterface` (see what I said about the optimization it brings to PSR-11 wrappers). The `Pimple\Container` API is not changed.

skalpa commented:
@fabpot I updated the C extension and fixed the typos/CS. The builds pass except the HHVM build, but that's due to #226.

- I had never seriously worked on a zend C extension before today, so hopefully it's not too bad
- I saw there were `.phpt` tests. Should I add some, even though the changes are covered by the PHPUnit tests ?
- Please send my thanks to jpauli, some of his articles helped make the experience more enjoyable than I expected :smile:

I'll handle your last comment once you help me understand it.

fabpot commented:
Some files have been converted to 0755 instead of 0644, that should be reverted.

fabpot commented on src/Pimple/Exception/UnknownIdentifierException.php:
```diff
@@ -0,0 +1,47 @@
+class UnknownIdentifierException extends \InvalidArgumentException implements NotFoundExceptionInterface
+{
+    /**
+     * Constructor.
```
Same here

fabpot commented on src/Pimple/Tests/PimpleTest.php:
```diff
@@ -115,6 +115,18 @@ public function testOffsetGetValidatesKeyIsPresent()
     }
 
+    /**
+     * @test
```
Instead of using the `@test` tag, you can name your test normally `testOfssetGet...` as the `@group legacy` annotation is enough to mark it as legacy.

fabpot commented:
You also need to bump the version to 3.1.

fabpot commented:
@jpauli Can you take some time to review the C extension changes? Thanks.

jpauli commented:
I would suggest such changes for consistency of the C API, and to fix compilation warnings :

https://github.com/jpauli/Pimple/commit/411dd06d1e73957c8861e9e31be60d9df2a79536

skalpa commented:
@jpauli I added the PHP methods declarations to the include file as you suggested, but I can't change the variables declarations to `static` as these variables are used from multiple files.

For instance, `pimple_ce_UnknownIdentifierException` is initialized in [pimple_exceptions.c](https://github.com/skalpa/Pimple/blob/9e58413846dd4705f6af2366f4de8ba1d577c295/ext/pimple/pimple_exceptions.c#L131) but is used by [pimple.c](https://github.com/skalpa/Pimple/blob/9e58413846dd4705f6af2366f4de8ba1d577c295/ext/pimple/pimple.c#L411-L414) afterwards.

If I declare the variable `static` from the include file:
- each `.c` file has its own `pimple_ce_UnknownIdentifierException` variable
- the one in `pimple.c` stays `NULL`
- the `pimple_throw_exception_string(pimple_ce_UnknownIdentifierException)` of `pimple.c#L412` segfaults (that happens when running the `phpt` tests)

AFAIK the only way I could make the variables static would be by merging all the files I wrote and putting everything in `pimple.c`.

Also, what compilation warnings do you get? I'm [not getting anything here](https://gist.github.com/skalpa/43a011c050a05f1fb05aa8efdfbcb0f3).

glensc commented:
@skalpa @jpauli somewhy replied in gist comment, not here:
https://gist.github.com/skalpa/43a011c050a05f1fb05aa8efdfbcb0f3#gistcomment-2116769

but you could also use `#include` trick to keep symbols in single file and use that "master" `.c` file for compilation:

`a.c`:
```
// a's implementation
```

`b.c`:
```
// b's implementation
```

`c.c`:
```
#include "a.c"
#include "b.c"
```

compilation
```
gcc c.c
```

glensc commented:
also, [visibility flags](https://gcc.gnu.org/wiki/Visibility) (doc [here](https://gcc.gnu.org/onlinedocs/gcc-4.7.1/gcc/Function-Attributes.html)) may help, but then the code would be gcc specific

[version map](https://gcc.gnu.org/wiki/SymbolVersioning) (doc [here](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Link-Options.html)) with `global:*` and `local:*` can also be used to control symbols visibility in resulting DSO.

skalpa commented:
@glensc Thanks a lot, I didn't get a notification when he commented on the gist. I was actually going to ping him again in the forthcoming days. I'll take care of it ASAP.

Thanks again 👍

stof commented on ext/pimple/pimple_psr.c:
```diff
@@ -0,0 +1,73 @@
+
+	INIT_NS_CLASS_ENTRY(tmp_ce_PsrContainerInterface,          PSR_CONTAINER_NS, "ContainerInterface",          pimple_ce_PsrContainerInterface_functions);
+	INIT_NS_CLASS_ENTRY(tmp_ce_PsrContainerExceptionInterface, PSR_CONTAINER_NS, "ContainerExceptionInterface", pimple_ce_PsrContainerExceptionInterface_functions);
+	INIT_NS_CLASS_ENTRY(tmp_ce_PsrNotFoundExceptionInterface,  PSR_CONTAINER_NS, "NotFoundExceptionInterface",  pimple_ce_PsrNotFoundExceptionInterface_functions);
```
Defining the PSR-11 interfaces inside ext-pimple looks weird to me. Isn't there a way to avoid it ?

jpauli replied:
The only way is to define them in a "psr extension", and have it as a dependency. 

So as of now, defining them in here is not a problem.
That could lead to conflicts with other extensions in the future though. A point that would then need further reflection.

stof replied:
yeah, not worth it indeed.
Best solution will be #228 anyway.

skalpa replied:
I honestly wondered about that too when I did it as it felt weird, but ended up thinking it was acceptable, as jpauli said.

There is [only one C extension on github](https://github.com/jbboehr/php-psr) that provides the PSR-11 interfaces, so basically we have the choice between:
- Implementing the interfaces in Pimple, effectively making the Pimple extension incompatible with the PSR extension
- Make the Pimple extension dependent on the PSR extension, which IMO can only be done if we tag a new major version.

Now, the point of this PR was precisely to help simplify/optimize the implementation of PSR-11 wrappers in a BC way that wouldn't require a new major release, so I went for the first choice, thinking there were very little chances somebody would complain about the conflict.

jpauli replied:
:+1: , let's not bother with that possibly huge subject now  :-)

skalpa commented:
@fabpot I'll merge the 3 `.c` files tonight, as jpauli suggested [here](https://gist.github.com/skalpa/43a011c050a05f1fb05aa8efdfbcb0f3). You'll be able to merge the PR after that.

fabpot commented:
Thank you @skalpa.
